compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,435 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+
+open Code
+open Stdlib
+
+let debug = Debug.find "globaldeadcode"
+
+let times = Debug.find "times"
+
+(** Definition of a variable [x]. *)
+type def =
+  | Expr of expr  (** [x] is defined by an expression. *)
+  | Param  (** [x] is a block or closure parameter. *)
+
+(** Liveness of a variable [x], forming a lattice structure. *)
+type live =
+  | Top  (** [x] is live and not a block. *)
+  | Live of IntSet.t  (** [x] is a live block with a (non-empty) set of live fields. *)
+  | Dead  (** [x] is dead. *)
+
+module G = Dgraph.Make_Imperative (Var) (Var.ISet) (Var.Tbl)
+
+module Domain = struct
+  type t = live
+
+  let equal l1 l2 =
+    match l1, l2 with
+    | Top, Top | Dead, Dead -> true
+    | Live f1, Live f2 -> IntSet.equal f1 f2
+    | _ -> false
+
+  let bot = Dead
+
+  (** Join the liveness according to lattice structure. *)
+  let join l1 l2 =
+    match l1, l2 with
+    | _, Top | Top, _ -> Top
+    | Live f1, Live f2 -> Live (IntSet.union f1 f2)
+    | Dead, Live f | Live f, Dead -> Live f
+    | Dead, Dead -> Dead
+end
+
+module Solver = G.Solver (Domain)
+
+let pure_expr pure_funs e = Pure_fun.pure_expr pure_funs e && Config.Flag.deadcode ()
+
+let definitions nv prog =
+  let defs = Array.make nv Param in
+  let set_def x d = defs.(Var.idx x) <- d in
+  Addr.Map.iter
+    (fun _ block ->
+      (* Add defs from block body *)
+      List.iter
+        ~f:(fun (i, _) ->
+          match i with
+          | Let (x, e) -> set_def x (Expr e)
+          | Assign (x, _) -> set_def x Param
+          | _ -> ())
+        block.body)
+    prog.blocks;
+  defs
+
+let variable_may_escape x (global_info : Global_flow.info) =
+  match global_info.info_variable_may_escape.(Var.idx x) with
+  | Escape | Escape_constant -> true
+  | No -> false
+
+(** Type of variable usage. *)
+type usage_kind =
+  | Compute  (** variable y is used to compute x *)
+  | Propagate  (** values of y propagate to x *)
+
+(** Compute the adjacency list for the dependency graph of given program. An edge between
+    variables [x] and [y] is marked [Compute] if [x] is used in the definition of [y]. It is marked
+    as [Propagate] if [x] is applied as a closure or block argument the parameter [y]. 
+    
+    We use information from global flow to try to add edges between function calls and their return values
+    at known call sites. *)
+let usages nv prog (global_info : Global_flow.info) : usage_kind Var.Map.t array =
+  let uses = Array.make nv Var.Map.empty in
```

Comment:
Shouldn't this use `Var.Tbl.make () Var.Map.empty` ?

Reply:
Yes and [646b216](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/646b216fade85c7567e0b559fedd661b1c972503) refactors to use Var.Tbl instead of Array in other places too